src/dune_rules/lib.ml

```diff
@@ -1116,7 +1139,85 @@ end = struct
+     | None -> Memo.return Status.Not_found
+     | Some db ->
+       let lib_id = Lib_info.lib_id info in
+       resolve_lib_id db lib_id)
+    >>= function
+    | Status.Found _ as x -> Memo.return x
+    | _ ->
+      let name = Lib_info.name info in
+      instantiate db name info ~hidden:(Some hidden)
+  ;;
+
+  let handle_resolve_result db ~super = function
+    | Ignore -> Memo.return Status.Ignore
+    | Redirect_in_the_same_db (_, name') -> find_internal db name'
+    | Redirect (db', lib_id') -> resolve_lib_id db' lib_id'
+    | Found info ->
+      let name = Lib_info.name info in
+      instantiate db name info ~hidden:None
+    | Invalid e -> Memo.return (Status.Invalid e)
+    | Not_found ->
+      (match db.parent with
+       | None -> Memo.return Status.Not_found
+       | Some db -> super db)
+    | Hidden { lib = info; reason = hidden; path = _ } -> resolve_hidden db ~info hidden
+  ;;
+
+  let handle_resolve_result_with_multiple_results db ~super = function
+    | Resolve_result r -> handle_resolve_result ~super db r
+    | Multiple_results candidates ->
+      let open Memo.O in
+      Nonempty_list.to_list candidates
+      |> Memo.List.filter_map ~f:(function
+        | Ignore -> Memo.return (Some Status.Ignore)
+        | Redirect_in_the_same_db (_, name') -> find_internal db name' >>| Option.some
+        | Redirect (db', lib_id') -> resolve_lib_id db' lib_id' >>| Option.some
+        | Found info ->
+          Lib_info.enabled info
+          >>= (function
+           | Disabled_because_of_enabled_if -> Memo.return None
+           | Normal | Optional ->
+             let name = Lib_info.name info in
+             instantiate db name info ~hidden:None >>| Option.some)
+        | Invalid e -> Memo.return (Some (Status.Invalid e))
+        | Not_found -> Memo.return None
+        | Hidden { lib = info; reason = hidden; path = _ } ->
+          resolve_hidden db ~info hidden >>| Option.some)
+      >>| (function
+       | [] -> Status.Not_found
+       | [ status ] -> status
+       | libs ->
+         List.fold_left libs ~init:Status.Not_found ~f:(fun acc status ->
+           match acc, status with
+           | Status.Found a, Status.Found b ->
+             let a_id = Lib_info.lib_id a.info in
+             let b_id = Lib_info.lib_id b.info in
+             (match Lib_id.equal a_id b_id with
+              | true -> acc
+              | false ->
+                let name =
+                  if Lib_name.equal a.name b.name then a.name else Lib_id.name a_id
+                and loc_a = Lib_info.loc a.info
+                and loc_b = Lib_info.loc b.info in
+                Status.Invalid (Error.duplicated ~loc_a ~loc_b ~name))
+           | Invalid _, _ -> acc
+           | (Found _ as lib), (Hidden _ | Ignore | Not_found | Invalid _)
+           | (Hidden _ | Ignore | Not_found), (Found _ as lib) -> lib
+           | (Hidden _ | Ignore | Not_found), (Hidden _ | Ignore | Not_found | Invalid _)
```

Comment:
This is rather hairy. Maybe try cleaning this up by pattern matching first on `acc` and then on `status`?

Reply:
I simplified it in 2f43f38dafebf2f2b8434901d909eef820254575